### PR TITLE
Subscript-crop `Raster` or `Vector` by bracket call `[]`, consistency and bug fixes

### LINKED ISSUE #213: Subscript a Raster with a Vector
**Is your feature request related to a problem? Please describe.**
The main principle of GeoUtils is to bridge Vectors and Rasters. Wouldn't it be cool if a Vector could be provided as a mask to a Raster by simply subscripting it?

**Describe the solution you'd like**

```python
img = gu.Raster("path/to/img.tif")
area_of_interest = gu.Vector("path/to/aoi.shp")


masked_img = img[area_of_interest]
```
The underyling functionality would be something alike:
```python
mask = area_of_interest.create_mask(img)

masked_img = img.copy()
masked_img.data.mask = masked_img.data.mask | mask.reshape(masked_img.data.shape)
```

**Describe alternatives you've considered**
The `masked_img` in the top could be several different types:

- A `Raster` in the same grid as the input
- A cropped `Raster` in a new grid.
- A 2D array of the masked values.
- A 1D array of the masked values.

I think a new `Raster` instance with the same grid should be the output, but I am open to discussion regarding this.

### LINKED ISSUE #262: `Raster.crop` does not reproject the bounding box of a Raster object provided as `cropGeom`
It's all in the title! I'm not sure why this behaviour would be intended?

We should create a polygon of the bounding box and reproject it to the same crs than `Raster` before cropping.

### LINKED ISSUE #268: Shapely triggers DeprecationWarnings
**Describe the bug**
Shapely 2.0 will be quite breaking, so there are lots of `DeprecationWarning`s floating around. This should be fixed before they turn into errors!

```
tests/test_geovector.py::TestSynthetic::test_generate_voronoi
  ./geoutils/geovector.py:448: ShapelyDeprecationWarning: Iteration over multi-part geometries is deprecated and will be removed in Shapely 2.0. Use the `geoms` property to access the constituent parts of a multi-part geometry.
    exteriors = [p.exterior for p in geom]
```

**To Reproduce**
Run the test suite.

**Expected behavior**
No (shapely) warnings should be triggered.

### LINKED ISSUE #276: Remove warning when using `Vector.create_mask()` only with `xres` argument
When creating a mask from a `Vector` only passing the resolution `xres` which behaves according to the description as _"Only xres is mandatory, by default yres=xres and bounds/crs are set to self's."_, a UserWarning is raised which might not desirable (or need to be rephrased):
```python
        mask = outlines.create_mask(xres=res)
```
>  UserWarning: Bounds not a multiple of xres/yres, use rounded bounds
>   warnings.warn("Bounds not a multiple of xres/yres, use rounded bounds")

This will always appear, as the shapefile bounds will virtually always not be rounded to a multiple of `xres`. So I'm thinking we could hide this warning in that occurence? Alternatively, we could rephrase as the current warning lets us think something went wrong (when it didn't), saying for instance "_Bounds were not a multiple of xres/yres, so lower bounds were selected to define the start of the grid and upper bounds redefined._"

### LINKED ISSUE #333: Polygonize does not support all `Raster` dtypes
Error in `geopandas` using `GeoDataFrame.from_features`:
```
ValueError: image dtype must be one of: int16, int32, uint8, uint16, float32
```

Would be more practical to be able to go directly from mask -> shapefile, or even other integer/float dtypes.

### PR DESCRIPTION
**Pull request by rhugonnet**
## Summary

This PR updates the `get_geoprojected_bounds` function, moving its core to `projtools` to make it consistent and accessible for both `Raster` and `Vector` classes. 
Then, it adds the same options in `Vector.crop()` than for `Raster.crop()` (= passing another `Raster`, `Vector` or a list of bounds). 
Finally, it implements the `__getitem__` (= `[]`) method for `Raster` and `Vector`, which calls their underlying `crop` functionality (= subsetting), as suggested previously by @erikmannerfelt.

```python
rst = Raster(fn_rst)
vec = Vector(fn_vec)

# Get only Raster on extent of Vector
rst = rst[vec]
```

Other than that, several bug fixes and test additions.

## Discussion on the `__getitem__` implementation

Several things:

- We could add the option of provides `slice` objects, which would then be used on `self.data` for `Raster` (subsetting `np.MaskedArray` in 2D) and `self.ds` on `Vector` (subsetting `gpd.GeoDataFrame` in 1D).
- We could, by default, `reproject` instead of `crop` when it is `Raster[Raster]` (only one where it makes sense)? The resampling algorithm could be passed as second argument `Raster[Raster, "bilinear"]`.

## Related issues and more details

This PR:
- Resolves #276,
- Resolves #268,
- Resolves #262,
- Resolves #213,
- Resolves #333.

Also it:
- Removes the dependency of `projtools.py` on `raster.py` and `geovector.py`,
- Moves the core of the `get_geoprojected_bounds` function in `Raster` to `projtools.py`,
- Adds `get_geoprojected_bounds` as a method for the `Vector` class,
- Mirrors the crop functionality of the `Raster` class for `Vector` objects (can provide any type of geometry: bounds, `Raster`, `Vector`),
- Adds more test for `Raster.polygonize()`,
- Solves a new bug with checking void in `intersection` triggered by package updates.

### CODE REVIEW

File: tests/test_georaster.py [hunk 1]
```diff
@@ -750,28 +771,42 @@ def test_crop(self, data: list[str]) -> None:
             cropGeom[2] - rand_float * r.res[0],
             cropGeom[3] - rand_float * abs(r.res[1]),
         ]
-        r_cropped = r.crop(cropGeom2, inplace=False, mode="match_extent")
+        with warnings.catch_warnings():
```
Comment (erikmannerfelt) [inline tests/test_georaster.py:774]: Could you add a comment for why UserWarnings are expected here?

File: tests/test_georaster.py [hunk 2]
```diff
@@ -704,6 +711,20 @@ def test_crop(self, data: list[str]) -> None:
         r_cropped2 = r.crop(r_cropped, inplace=False)
         assert r_cropped2 == r_cropped
 
+        # Check that bound reprojection is done automatically if the CRS differ
+        with warnings.catch_warnings():
```
Comment (erikmannerfelt) [inline tests/test_georaster.py:715]: Same as below; why do we expect warnings here?

File: tests/test_georaster.py (hunk 1 above)
Comment (erikmannerfelt) [reply tests/test_georaster.py:774]: Also, I'm pretty sure that all our warnings are UserWarnings, so that means all of them get suppressed.

File: geoutils/spatial_tools.py [hunk 3]
```diff
@@ -136,8 +137,8 @@ def load_multiple_rasters(
 
     # Optionally, crop the rasters
     if crop:
-        # Check that intersection is not void
-        if intersection == ():
+        # Check that intersection is not void (changed to NaN instead of empty tuple end 2022)
+        if intersection == () or all(math.isnan(i) for i in intersection):
```
Comment (erikmannerfelt) [inline geoutils/spatial_tools.py:141]: What's the advantage of  `math.isnan` to `~np.isfinite`?

File: tests/test_georaster.py [hunk 4]
```diff
@@ -1798,22 +1836,42 @@ def test_resampling_str(self) -> None:
...
-        value = np.unique(img)[0]
+        img = gr.Raster(example)
 
-        pixel_area = np.sum(img == value) * img.res[0] * img.res[1]
+        # -- Test 1: basic functioning of polygonize --
 
-        polygonized = img.polygonize(value)
+        # Get unique value for image and the corresponding area
+        value = np.unique(img)[0]
+        pixel_area = np.sum(img == value) * img.res[0] * img.res[1]
```
Comment (erikmannerfelt) [inline tests/test_georaster.py:1849]: Could you change this to `np.count_nonzero()`? I know they're synonymous in the case of booleans, but "count nonzero" makes things easier to understand in the long term in my opinion.

File: tests/test_georaster.py [hunk 5]
```diff
@@ -1798,22 +1836,42 @@ def test_resampling_str(self) -> None:
...
         assert polygon_area == pytest.approx(pixel_area)
         assert isinstance(polygonized, gv.Vector)
         assert polygonized.crs == img.crs
 
+        # -- Test 2: data types --
+
+        # Check that polygonize works as expected for any input dtype (e.g. float64 being not supported by GeoPandas)
+        for dtype in ["uint8", "int8", "uint16", "int16", "uint32", "int32", "float32", "float64"]:
+            img_dtype = img.copy()
+            with warnings.catch_warnings():
```
Comment (erikmannerfelt) [inline tests/test_georaster.py:1865]: What warnings are suppressed?

**Review by erikmannerfelt — COMMENTED** [review]
Hey Romain!
These additions are awesome. You're really contributing to this being maintainable in the long term! On the subject of maintainability, I have some comments on details that I struggle to understand, but generally, it's great!

Regarding your two suggestions about the subscripting:
- On the first, do you mean to slice by indices like `arr.isel()` or `df.iloc[]` in xarray and pandas, respectively? If so, it would actually be pretty cool, but I struggle to see a use case. For iterating along blocks perhaps? Or subsetting for debug purposes? Those two are good arguments in my head, but for the amount of complexity they introduce, it might not be a great idea unless there are more uses.
- I personally think the subscripting syntax should only use crop. Reproject can be very cpu-intensive and should in my opinion always be explicit. Cropping, however, is really fast, so there the subscripting syntax makes a ton of sense to me.

File: geoutils/projtools.py [hunk 6]
```diff
@@ -15,12 +15,9 @@
 from shapely.geometry.base import BaseGeometry
 from shapely.geometry.polygon import Polygon
 
-from geoutils.georaster import Raster
-from geoutils.geovector import Vector
-
 
 def bounds2poly(
-    boundsGeom: list[float] | rio.io.DatasetReader | Raster | Vector,
+    boundsGeom: list[float] | rio.io.DatasetReader,
```
Comment (adehecq) [inline geoutils/projtools.py:20]: why did you delete it? It's actually meant to take a Raster or Vector instance.

File: geoutils/georaster/raster.py [hunk 7]
```diff
@@ -1375,13 +1385,13 @@ def crop(
     def reproject(
         self: RasterType,
         dst_ref: RasterType | rio.io.Dataset | str | None = None,
-        dst_crs: CRS | str | None = None,
+        dst_crs: CRS | str | int | None = None,
```
Comment (adehecq) [inline geoutils/georaster/raster.py:1388]: why is int needed? For EPSG codes? If so, it should be clarified in the docstring what argument can be passed.

File: tests/test_georaster.py [hunk 8]
```diff
@@ -750,28 +771,42 @@ def test_crop(self, data: list[str]) -> None:
...
+            r_cropped2 = r.crop(r_cropped, inplace=False, mode="match_extent")
         assert r_cropped2 == r_cropped
 
         # -- Test with CropGeom being a Vector -- #
         outlines = gu.Vector(outlines_path)
-        outlines.ds = outlines.ds.to_crs(r.crs)
-        r_cropped = r.crop(outlines, inplace=False)
+
+        # First, we reproject manually the outline
+        outlines_reproj = gu.Vector(outlines.ds.to_crs(r.crs))
```
Comment (adehecq) [inline tests/test_georaster.py:792]: When reading your PR, I was thinking that in the same effort (making crop be similar between Raster and Vector), we should make reproject more consistent between the two, i.e. have a Vector.reproject function.

**Review by adehecq — APPROVED** [review]
Great idea to make crop more consistent and thanks for fixing all the small issues.
The brackets are a nice feature, but they would need to be documented at some point, otherwise, we'll just forget that it's possible !

File: geoutils/projtools.py (hunk 6 above)
Comment (rhugonnet) [reply geoutils/projtools.py:20]: Cannot use `Raster` or `Vector` in `projtools` due to circular import

File: tests/test_georaster.py (hunk 1 above)
Comment (rhugonnet) [reply tests/test_georaster.py:774]: Done, added a comment everywhere + the `message` argument with a regex pattern to only catch the desired warnings.

File: tests/test_georaster.py (hunk 2 above)
Comment (rhugonnet) [reply tests/test_georaster.py:715]: Done as above

File: geoutils/spatial_tools.py (hunk 3 above)
Comment (rhugonnet) [reply geoutils/spatial_tools.py:141]: Hmmm, curious, for some reason when I tried with `np.isnan` the first time, it did not work, because the output of geopandas is a `float('nan')`, but now it works element-wise...
I changed to `np.isnan`.

File: tests/test_georaster.py (hunk 5 above)
Comment (rhugonnet) [reply tests/test_georaster.py:1865]: Specified the match message.

File: tests/test_georaster.py (hunk 4 above)
Comment (rhugonnet) [reply tests/test_georaster.py:1849]: Done

File: geoutils/georaster/raster.py (hunk 7 above)
Comment (rhugonnet) [reply geoutils/georaster/raster.py:1388]: Done

File: geoutils/geovector.py [hunk 9]
```diff
@@ -120,7 +156,7 @@ def create_mask(
     ) -> np.ndarray:
         """
         Rasterize the vector features into a boolean raster which has the extent/dimensions of \
-the provided raster file.
+        the provided raster file.
```
[the remark this answers is not in the document]
Comment (rhugonnet) [reply geoutils/geovector.py:159]: I don't know either... except that it makes the code hard to read, and does not respect Python indentation. I haven't seen that in any of the other packages I went through (NumPy, Pandas, etc).
I'll open an issue on this. For now, I know it displays fine in the API with this kind of tabulated text.

[the remark this answers is not in the document]
Comment (rhugonnet) [reply geoutils/geovector.py:159]: Opened #338

File: tests/test_georaster.py (hunk 8 above)
Comment (rhugonnet) [reply tests/test_georaster.py:792]: I was thinking the same, but then reproject would only be about the CRS, right?

**Comment by rhugonnet** [comment]
Thanks a lot! All updated, merging :smile:

File: tests/test_georaster.py (hunk 8 above)
Comment (adehecq) [reply tests/test_georaster.py:802]: Yes that's right, but still it would be more consistent between the two classes.

Comment (adehecq) [reply tests/test_georaster.py:802]: Just added issue #340
